test/scripts/e2e_subs/rest.sh

```diff
@@ -35,21 +35,242 @@ function base_call {
 }
 
 
+function base_post_call {
+  curl -X POST --data-binary @/${TEMPDIR}/$4 -o "$3" -w "%{http_code}" -q -s -H "Authorization: Bearer $1" "$NET$2"
+}
+
+
+function base_delete_call {
+  curl -X DELETE -o "$3" -w "%{http_code}" -q -s -H "Authorization: Bearer $1" "$NET$2"
+}
+
 function call_admin {
   base_call "$ADMIN_TOKEN" "$1" "$2"
 }
 
+function call_post_admin {
+  base_post_call "$ADMIN_TOKEN" "$1" "$2" "$3"
+}
+
+function call_delete_admin {
+  base_delete_call "$ADMIN_TOKEN" "$1" "$2" "$3"
+}
 
 function call {
   base_call "$PUB_TOKEN" "$1" "$2"
 }
 
+function call_post {
+  base_post_call "$PUB_TOKEN" "$1" "$2"
+}
+
+function call_delete {
+  base_delete_call "$PUB_TOKEN" "$1" "$2"
+}
+
 
 function fail_and_exit {
   printf "\n\nFailed test - $1 ($2): $3\n\n"
   exit 1
 }
 
+# $1 - test description.
+# $2 - query
+# $3 - expected status code
+# $4 - the file to upload
+# $5... - substring that should be in the response
+function call_post_and_verify {
+   local DESCRIPTION="$1"
+   shift
+   local QUERY="$1"
+   shift
+   local EXPECTED_CODE="$1"
+   shift
+   local FILENAME_TO_UPLOAD="$1"
+   shift
+
+   echo "MATCHING $@"
+   curl_post_test "$DESCRIPTION" "$QUERY" "$EXPECTED_CODE" true "$FILENAME_TO_UPLOAD" "$@"
+}
+
+
+# CURL POST Test - POST query and verify results
+# $1 - test description.
+# $2 - query
+# $3 - expected status code
+# $4 - match result
+# $5 - the file to upload
+# $6... - substring(s) that should be in the response
+function curl_post_test {
+  local DESCRIPTION="$1"
+  shift
+  local QUERY="$1"
+  shift
+  local EXPECTED_CODE="$1"
+  shift
+  local MATCH_RESULT="$1"
+  shift
+  local FILENAME_TO_UPLOAD="$1"
+  shift
+
+
+  local SUBSTRING
+
+  local START=$SECONDS
+
+  set +e
+  local CODE
+  if [[ "$USE_ADMIN" = true ]]; then
+    CODE=$(call_post_admin "$QUERY" "${TEMPDIR}/curl_out.txt" "$FILENAME_TO_UPLOAD")
+  else
+    CODE=$(call_post "$QUERY" "${TEMPDIR}/curl_out.txt" "$FILENAME_TO_UPLOAD")
+  fi
```

Comment:
I think we should try to avoid having 4 copies of this verify logic. One "nice" way to do that might be to put the post/delete/get option at parameter 1 and have a common `verify` function:
```
function curl_post_test {
  local FILENAME_TO_UPLOAD="$1"
  shift

  local START=$SECONDS

  set +e
  local CODE
  if [[ "$USE_ADMIN" = true ]]; then
    CODE=$(call_post_admin "$QUERY" "${TEMPDIR}/curl_out.txt" "$FILENAME_TO_UPLOAD")
  else
    CODE=$(call_post "$QUERY" "${TEMPDIR}/curl_out.txt" "$FILENAME_TO_UPLOAD")
  fi

  verify $? "$CODE" "$@"
}

function verify {
  local SUCCESS=$1
  shift
  local CODE=$1
  shift
  local DESCRIPTION="$1"
  shift
  local QUERY="$1"
  shift
  local EXPECTED_CODE="$1"
  shift
  local MATCH_RESULT="$1"
  shift

  if [[ $SUCCESS != 0 ]]; then
    cat $CURL_TEMPFILE
    fail_and_exit "$DESCRIPTION" "$QUERY" "curl had a non-zero exit code."
  fi
  set -e

  RES=$(cat "${TEMPDIR}/curl_out.txt")
  if [[ "$CODE" != "$EXPECTED_CODE" ]]; then
    fail_and_exit "$DESCRIPTION" "$QUERY" "unexpected HTTP status code expected $EXPECTED_CODE (actual $CODE): $RES"
  fi

  ... etc ...
}
```